Json api compatibility

alexshook opened the pull request.

alexshook committed:
WIP Add JSON API media type, modify identifying a list

alexshook committed:
Fix paths to fit API Blueprint syntax

- We need {} around parameters in the API Blueprint markdown doc.
- Also allows a `named` parameter describing the action to identify
endpoints. It's chiefly used to identify endpoints that return a
collection.
- If no `named` parameter is given, it falls back to the legacy method
of identifying by parsing the route for `:id`.

alexshook committed:
WIP

alexshook committed:
Add sample example for widget with a named param

alexshook committed:
Add sample example endpoint for nested route

alexshook committed:
Simplify route formatting

- Adds a method to API Blueprintify routes. API Blueprint requires route
ids to be wrapped with {} instead of the convention of using a : to
denote an id.
- We no longer have to check to see if a route is a collection route now
either.

alexshook committed:
Use new API Blueprint method for route path

alexshook committed:
Fix typo :see_no_evil:

alexshook committed:
Add sample entity for testing

- Currently an empty class but we may need to fill it in later.

alexshook committed:
Use detail hash to pass information instead of named

alexshook committed:
Try to use the entity for the SampleGenerator

- If the entity hasn't been passed from the API in detail or something
goes wrong, it will fall back to the legacy way of getting the response
body.

alexshook committed:
Don't depend on sending the entity from the api

- This does a bit of hard coding but elimiates the need for sending back
the name of the entity in the detail section of the endpoint
documentation.
- Doesn't fail anything but needs some tests still.

alexshook committed:
Remove detail from example

We are sending it to the gem in a different way.

alexshook committed:
Consolidate example classes used for tests

alexshook committed:
Use the documentation hash to send the endpoint action

alexshook committed:
Generate JSON API entity response when possible

If not possible, uses the legacy method of generating the hash.

alexshook committed:
Add test for generating sample with JSON API entity

- I'm not really pleased wtih this.
- My attempt to make this happen with null object didn't work out like I
thought it would.
- Short of implementing the API and grape jsonapi entity gem here, this
test is about as good as I can make it.

alexshook committed:
Fix some typos in tests

alexshook committed:
Remove unneccessary method

Turns out we don't need path_without_format to do anything else so this
can go back as it was.

alexshook committed:
Rubocop fixes

alexshook commented on spec/grape-apiary/sample_generator_spec.rb:
```diff
@@ -36,4 +36,24 @@
+    context 'when the JSON API entity exists' do
+      let(:entity) { V20170505::Entities::Widget }
+
+      it 'is the entity hash' do
```
I'm not happy with this but my null object solution failed and I'm not going to reimplement the API and the grape jsonapi entities gem in this gem to test this method.

alexshook committed:
It's match, not match?

inactdev commented on lib/grape-apiary/sample_generator.rb:
```diff
@@ -2,26 +2,17 @@ module GrapeApiary
-
-      hash
+      generate_jsonapi_entity_sample_hash
+    rescue
```
I think you should specify the error to rescue from here.  I like the way you did this tho ! 👌

alexshook replied:
heh it could be any number of errors, which is why the blanket rescue from StandardError and the fall back to the legacy generator. I'll try to narrow them down, there may only be three or four possible errors.

inactdev commented on spec/grape-apiary/route_spec.rb:
```diff
@@ -7,15 +7,70 @@
-    expect(subject.path_without_format).to eq('/widgets')
+  describe '#api_blueprintified_path' do
+    context 'when the route is GET /widgets/:id' do
+      subject(:individual_widget) { routes.second }
```
Is there a little less brittle/more obvious way for you to test the different route types ?  `routes.#{number}` isn't super discoverable 😕

alexshook replied:
Sadly, no... this is how the whole gem works, as in it loads in all the routes and then operates on each one. If you can think of a better way halp plz open a pull request.

inactdev reviewed (changes requested):
Great job on this.  Just a couple comments and hey, if it works....  #YOLO

alexshook committed:
Add Jsonapi formatter for top level included member

- Stubs Grape::Jsonapi::Formatter in tests due to a namespacing issue
with Grape::Formatter.
- For some reason stub_const will only work inside the it block, idk
why.
- Also fixes spacing in other sample test.

alexshook committed:
Fix SampleGenerator tests

alexshook committed:
Merge remote-tracking branch 'upstream/grape-1.0' into json_api_compatibility
